com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/PollExecuteTaskStatusExecution.java

```diff
@@ -0,0 +1,122 @@
+            this.app = StepsUtil.getApp(execution.getContext());
+        }
+
+        public AsyncExecutionState execute() {
+            CloudTask.State currentState = getCurrentState();
+            reportCurrentState(currentState);
+            saveAppLogs();
+            return handleCurrentState(currentState);
+        }
+
+        private CloudTask.State getCurrentState() {
+            ClientExtensions clientExtensions = execution.getClientExtensions();
+            List<CloudTask> allTasksForApp = clientExtensions.getTasks(app.getName());
+
+            return findTaskWithGuid(allTasksForApp, taskToPoll.getMeta().getGuid()).getState();
+        }
+
+        private CloudTask findTaskWithGuid(List<CloudTask> allTasksForApp, UUID guid) {
+            return allTasksForApp.stream().filter(task -> task.getMeta().getGuid().equals(guid)).findAny().orElseThrow(
+                () -> new IllegalStateException(format(Messages.COULD_NOT_FIND_TASK_WITH_GUID, guid)));
+        }
+
+        private void reportCurrentState(CloudTask.State currentState) {
+            execution.getStepLogger().info(Messages.TASK_EXECUTION_STATUS, currentState.toString().toLowerCase());
+        }
+
+        private void saveAppLogs() {
+            CloudFoundryOperations client = execution.getCloudFoundryClient();
+            StepsUtil.saveAppLogs(execution.getContext(), client, recentLogsRetriever, app, LOGGER,
+                execution.getProcessLoggerProviderFactory());
+        }
+
+        private AsyncExecutionState handleCurrentState(CloudTask.State currentState) {
+            if (isFinalState(currentState)) {
+                return handleFinalState(currentState);
+            }
+            return checkTimeout();
+        }
+
+        private boolean isFinalState(CloudTask.State currentState) {
+            return currentState.equals(CloudTask.State.FAILED) || currentState.equals(CloudTask.State.SUCCEEDED);
+        }
+
+        private AsyncExecutionState handleFinalState(CloudTask.State state) {
+            if (state.equals(CloudTask.State.FAILED)) {
+                execution.getStepLogger().error(Messages.ERROR_EXECUTING_TASK_ON_APP, taskToPoll.getName(), app.getName());
+                StepsUtil.setStepPhase(execution, StepPhase.RETRY);
+                return AsyncExecutionState.ERROR;
+            }
+            StepsUtil.setStepPhase(execution, StepPhase.EXECUTE);
+            return AsyncExecutionState.FINISHED;
+        }
+
+        private AsyncExecutionState checkTimeout() {
+            if (StepsUtil.hasTimedOut(execution.getContext(), currentTimeSupplier)) {
+                String message = format(Messages.EXECUTING_TASK_ON_APP_TIMED_OUT, taskToPoll.getName(), app.getName());
+                execution.getStepLogger().error(message);
+                StepsUtil.setStepPhase(execution, StepPhase.RETRY);
+                return AsyncExecutionState.ERROR;
+            }
+            StepsUtil.setStepPhase(execution, StepPhase.POLL);
```

Comment:
What will happen if some developer forgets to add such a line in his AsyncExecution?

Reply:
It is not important to set the phase here... it is leftover :)